### PR TITLE
npy serialization and checkpoints

### PR DESCRIPTION
**Pull request by ry**
Depends on #287

### CODE REVIEW

**Comment by ry** [comment]
@piscisaureus PTAL.

File: src/disk_experiment_test.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  const checkpointDir = path.join(expDir, checkpointDirs[0]);
+  assert(isDir(checkpointDir));
+  assert(fs.existsSync(path.join(checkpointDir, "hello/world.npy")));
+
+  // Try to load the checkpoint we just saved.
+  const exp_ = new DiskExperiment("exp1");
+  await exp_.createOrRestore();
+  assert(exp_.params.length === 1);
+  const t = exp_.params.get("hello/world");
+  assertAllEqual(t, [1, 2, 3]);
```
Comment (ry) [inline src/disk_experiment_test.ts]: The windows tests are failing here. Need windows to debug.

File: src/npy.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  // First figure out how long the file is going to be so we can create the
+  // output ArrayBuffer.
+  const magicStr = "NUMPY";
+  const versionStr = "\x01\x00";
+  const [ d, fo, s ] = [ descr, "False", tensor.shape.join(",") + "," ];
+  let header = `{'descr': '${d}', 'fortran_order': ${fo}, 'shape': (${s}), }`;
+  const unpaddedLength = 1 + magicStr.length + versionStr.length +
+                         2 +  header.length;
+  // Spaces to 16-bit align.
+  const padding = " ".repeat(16 - unpaddedLength % 16);
```
Comment (piscisaureus) [inline src/npy.ts]: This inserts 16 spaces when 0 are needed.

File: src/disk_experiment.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+export class DiskExperiment extends Experiment {
+  constructor(readonly name: string, opts?: ExperimentOpts) {
+    super(name, opts);
+    if (this.opts.saveOnExit) {
+      process.on("exit", async(exitCode) => {
+        // If there's no error, save the checkpoint one file last time.
+        if (exitCode === 0 && this.step > 0) {
+          await this.save();
```
Comment (piscisaureus) [inline src/disk_experiment.ts]: You can't call async functions in the 'exit' handler.
Use 'beforeExit' if you want to auto-save when node is about to exit "naturally".
To auto-save when process.exit() is called, we need a `.saveSync()` method.

File: src/disk_experiment.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    rimraf.sync(p);
+    await print("Checkpoint deleted", p);
+  }
+
+  private checkpointPath(step: number): string {
+    return path.join(this.dir, String(step).padStart(8, "0")) + "/";
+  }
+
+  // Saves a checkpoint to disk.
+  async save(): Promise<void> {
```
Comment (piscisaureus) [inline src/disk_experiment.ts]: Why is this function async? All fs calls inside are sync...

File: src/npy.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  }[tensor.dtype];
+
+  // First figure out how long the file is going to be so we can create the
+  // output ArrayBuffer.
+  const magicStr = "NUMPY";
+  const versionStr = "\x01\x00";
+  const [ d, fo, s ] = [ descr, "False", tensor.shape.join(",") + "," ];
+  let header = `{'descr': '${d}', 'fortran_order': ${fo}, 'shape': (${s}), }`;
+  const unpaddedLength = 1 + magicStr.length + versionStr.length +
+                         2 +  header.length;
```
Comment (piscisaureus) [inline src/npy.ts]: space

File: src/disk_experiment.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      fs.writeFileSync(tensorPath, new Buffer(ab));
+    }
+    const megs = (totalSize / (1024 * 1024)).toFixed(2);
+    await print(`Checkpoint saved ${megs}M`, checkpointPath);
+  }
+}
+
+export function isDir(p: string): boolean {
+  try {
+    return fs.lstatSync(p).isDirectory();
```
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: No reason in particular. Are you suggesting I use `statSync` instead?

File: src/disk_experiment.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+import * as npy from "./npy";
+import { Params, params as createParams } from "./params";
+import { assert } from "./util";
+
+/** Returns "$HOME/.propel/" or PROPEL_DIR env var. */
+function propelDir(): string {
+  if (process.env.PROPEL_DIR) {
+    return process.env.PROPEL_DIR;
+  } else {
+    return path.join(process.env.HOME, ".propel/");
```
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: Done

File: src/disk_experiment.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    this.step_ = step;
+    this.currentParams = params;
+    return params;
+  }
+
+  async checkpoints(): Promise<number[]> {
+    const out = [];
+    for (const fn of fs.readdirSync(this.dir)) {
+      if (fn.match(/^\d+$/)) {
+        out.push(Number(fn));
```
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: Done

File: src/disk_experiment.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    let totalSize = 0;
+    for (const [name, tensor] of this.currentParams) {
+      const tensorPath = path.join(checkpointPath, name) + ".npy";
+      ensureDirExists(path.dirname(tensorPath));
+      const ab = await npy.serialize(tensor);
+      totalSize += ab.byteLength;
+      fs.writeFileSync(tensorPath, new Buffer(ab));
+    }
+    const megs = (totalSize / (1024 * 1024)).toFixed(2);
+    await print(`Checkpoint saved ${megs}M`, checkpointPath);
```
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: Done

File: src/npy.ts (hunk 2 above)
Comment (ry) [reply src/npy.ts]: Fixed.

File: src/npy.ts (hunk 5 above)
Comment (ry) [reply src/npy.ts]: Fixed.

File: src/disk_experiment.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: Changed to statSync.

File: src/disk_experiment_test.ts (hunk 1 above)
Comment (ry) [reply src/disk_experiment_test.ts]: Fixed.

File: src/disk_experiment.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  async restore(step: number): Promise<Params> {
+    const p = this.checkpointPath(step);
+    const npyFiles = filePatternSearch(p, /\.npy$/);
+    const params = createParams();
+    for (const fn of npyFiles) {
+      const name = fn.replace(p, "").replace(/\.npy$/, "");
+      const tensor = await npy.load(fn);
+      params.set(name, tensor);
```
[the remark this answers is not in the document]
Comment (ry) [reply src/disk_experiment.ts]: Fixed.

**Pull request merged**